File: content-solutions/heap/fewest-refuels.java
import java.util.*;

class Solution {
    public int solution(int goal, int start, int[][] springs) {
        PriorityQueue<Integer> h = new PriorityQueue<>(Collections.reverseOrder());
        long fuel = start;
        int stops = 0;
        for (int i = 0; i <= springs.length; i++) {
            long pos = i < springs.length ? springs[i][0] : goal;
            while (fuel < pos) {
                if (h.isEmpty()) return -1;
                fuel += h.poll();
                stops++;
            }
            if (i < springs.length) h.offer(springs[i][1]);
        }
        return stops;
    }
}

File: content-solutions/heap/fewest-refuels.py
import heapq


def solution(goal, start, springs):
    fuel = start
    stops = 0
    h = []
    for pos, water in springs + [[goal, 0]]:
        while fuel < pos:
            if not h:
                return -1
            fuel += -heapq.heappop(h)
            stops += 1
        heapq.heappush(h, -water)
    return stops
